File: src/edu/uf/interactable/Leukocyte.cpp
/*
 * Leukocyte.cpp
 *
 *  Created on: Jul 17, 2024
 *      Author: henriquedeassis
 */
#include "../intracellularState/IntracellularModel.h"
#include "../utils/Constants.h"
#include "../utils/Rand.h"
#include "Leukocyte.h"
#include "../compartments/GridFactory.h"
#include "../compartments/Voxel.h"
#include "../utils/Util.h"
#include "PositionalAgent.h"

namespace edu {
namespace uf {
namespace interactable {

using namespace edu::uf::compartments;
using namespace edu::uf::utils;

Leukocyte::Leukocyte(intracellularState::IntracellularModel* network)
    : Leukocyte(0.0, network) {}

Leukocyte::Leukocyte(double ironPool, intracellularState::IntracellularModel* network)
    : Cell(network), engaged(false) {
    setIronPool(ironPool);
}

bool Leukocyte::isEngaged() const {
    return engaged;
}

void Leukocyte::setEngaged(bool engaged) {
    this->engaged = engaged;
}

void Leukocyte::kill() {
    if(Rand::getRand().randunif() < utils::constexprants::PR_KILL) {
        for(auto agent : phagosome) {
            incIronPool(agent->getIronPool());
            agent->incIronPool(-agent->getIronPool());
            agent->die();
        }
        phagosome.clear();
    }
}

void Leukocyte::setPhagosome(std::vector<Cell*>& phagosome) {
	this->phagosome = phagosome;
}

std::vector<Cell*>& Leukocyte::getPhagosome() {
	return this->phagosome;
}

void Leukocyte::addAspergillus(Cell* aspergillus) {
	phagosome.push_back(aspergillus);
}

void Leukocyte::move(int x, int y, int z, int steps) {
    if(steps < getMaxMoveSteps()) {
        auto grid = GridFactory::getGrid();
        Voxel* oldVoxel = grid[x][y][z];
        Util::calcDriftProbability(oldVoxel, this);
        Voxel* newVoxel = Util::getVoxel(oldVoxel, Rand::getRand().randunif());

        oldVoxel->removeCell(getId());
        newVoxel->setCell(this);
        steps += 1;

        for(auto agent : phagosome) {
            if(auto positionalAgent = dynamic_cast<PositionalAgent*>(agent)) {
                positionalAgent->setX(newVoxel->getX() + Rand::getRand().randunif());
                positionalAgent->setY(newVoxel->getY() + Rand::getRand().randunif());
                positionalAgent->setZ(newVoxel->getZ() + Rand::getRand().randunif());
            }
        }
        move(newVoxel->getX(), newVoxel->getY(), newVoxel->getZ(), steps);
    }
}

} /* namespace interactable */
} /* namespace uf */
} /* namespace edu */

File: java/src/edu/uf/interactable/Leukocyte.java
package edu.uf.interactable;

import java.util.ArrayList;
import java.util.List;

import edu.uf.compartments.GridFactory;
import edu.uf.compartments.Voxel;
import edu.uf.interactable.Afumigatus.Afumigatus;
import edu.uf.intracellularState.IntracellularModel;
import edu.uf.utils.Constants;
import edu.uf.utils.Rand;
import edu.uf.utils.Util;

public abstract class Leukocyte extends Cell{

    //public static final int FREE = 0;
    //public static final int INTERACTING = 1;

    
    
    //private Phagosome phagosome;
    private List<InfectiousAgent> phagosome;
    
    private boolean engaged;
    
    //private int state;

	public Leukocyte(IntracellularModel network) {
    	this(0.0, network);
    }
    
    public Leukocyte (double ironPool, IntracellularModel network) {
    	super(network);
        this.setIronPool(ironPool);
        //this.setStatusIteration(0);
        //this.setStateIteration(0);
        //this.phagosome = new Phagosome();
        this.phagosome = new ArrayList<>();
    }
	
	/*public int getState() {
		return state;
	}

	public void setState(int state) {
		this.state = state;
	}*/

	public void setPhagosome(List<InfectiousAgent> phagosome) {
		this.phagosome = phagosome;
	}

    public abstract int getMaxCell();
    

    public List<InfectiousAgent> getPhagosome() {
        return this.phagosome; 
    }

	public boolean isEngaged() {
		return engaged; 
	}

	public void setEngaged(boolean engaged) {
		this.engaged = engaged;
	}

	/**
	 * <strong>This method is too specific. It should be changed.</strong>
	 * @param aspergillus
	 */
    public void addAspergillus(Afumigatus aspergillus) {
        this.phagosome.add(aspergillus);
    }

    public void kill() {
        if(Rand.getRand().randunif() < Constants.PR_KILL) {
        	for(InfectiousAgent a : this.phagosome) {
        		//System.out.print("BLA\t");
                this.incIronPool(a.getIronPool());
                a.incIronPool(-a.getIronPool());
                a.die();
                //System.out.println();
        	}
            this.phagosome.clear();
        }
    }

    public void move(int x, int y, int z, int steps) {
        if(steps < this.getMaxMoveSteps()) {
        	
        	Voxel oldVoxel = GridFactory.getGrid()[x][y][z];
        	Util.calcDriftProbability(oldVoxel, this);
            Voxel newVoxel = Util.getVoxel(oldVoxel, Rand.getRand().randunif());
        	
            oldVoxel.removeCell(this.getId());
            //System.out.println(oldVoxel + " - " + newVoxel + " " + this);
            newVoxel.setCell(this);
            steps = steps + 1;
            PositionalAgent p = null;
            for(InfectiousAgent a : this.phagosome) {
            	if(a instanceof PositionalAgent) {
            		p = (PositionalAgent) a;
            		p.setX(newVoxel.getX() + Rand.getRand().randunif());
            		p.setY(newVoxel.getY() + Rand.getRand().randunif());
            		p.setZ(newVoxel.getZ() + Rand.getRand().randunif());
            	}
            }
            move(newVoxel.getX(), newVoxel.getY(), newVoxel.getZ(), steps);
        }
    }
    
	
}
